### ISSUE TITLE
Rethinking the attachment system for org-roam v2?

### ISSUE THREAD

**Issue by archiif** [hlissner/doom-emacs#5259]
### Describe your request

The default Doom's attachment system for org-mode works well for org-roam v1 because it assumes one top level ID per file that probably won't change. 

But in org-roam v2, creating ID links to specific headings will become a much more frequent activity. The problem is when the attachments on the heading were made before the ID to that heading was initialized (thus the file would be located relative to the buffer-level ID). This would cause links to those attachments to be broken because org-mode would assume that the file is located relative to that heading ID.

To anticipate this, the default configuration for attachments should probably be adjusted.

### Briefly explain its use-case

1. Upgrade to org-roam v2
2. Use Doom's org attachment system without suffering link breakage when linking to headings

**Comment by Wetlize** [comment]
Doom uses the built-in attachment system from org-mode called `org-attach`. The behavior that you can observe comes from `org-attach-use-inheritance` which Doom sets to `t`. If you disable the attachment inheritance (by setting this variable to `nil`) it will do what you want.

**Comment by archiif** [comment]
Thanks for the tip, I tried disabling `org-attach-use-inheritance` and it works pretty nicely so far. But there's one quirk that I don't like about it, which is that it prevents you from making attachments as headings. I guess it's not exactly a standard workflow, but sometimes I like using images as headings.

When inheritance is disabled, each heading needs to have an ID property defined before the attachment is defined, but when the attachment is located on the heading, org-attach is unable to associate the attachment with any ID (because the ID would be defined after it), thus the link breaks.

Though you're right that this probably should be relegated to each user's private config, but any tips to overcome the above quirk would be welcome.

**Comment by Wetlize** [comment]
> When inheritance is disabled, each heading needs to have an ID property defined before the attachment is defined, but when the attachment is located on the heading, org-attach is unable to associate the attachment with any ID (because the ID would have to be defined after it), thus the link breaks.

Does adding the next lines to your config help to mitigate this issue?
```emacs-lisp
(setq-hook! 'org-roam-find-file-hook
    ;; alternatively try `t` as the value
    org-id-link-to-org-use-id 'create-if-interactive)
```

**Comment by archiif** [comment]
Ah, I was just being stupid and completely misunderstood how `org-attach` works apparently. 

The only reason it behaved the way I described above is because the Doom's default configuration of `org-download` (which is what I happened to use to test the attachment system) inserts a newline after `[[attachment:...]]`.

So when you call a function from `org-download` on a heading, there will be a newline inserted between the heading (containing the attachment link) and the property drawer. This is apparently considered as an invalid property syntax, which is what broke the link.

Customizing `org-download-link-format-function` and removing the newline fixed this issue.

For now, I'm closing the issue since this seems satisfactory.
